File: tests/insight_metrics.py
"""Executable metric contract for Starling local Insight events (E28).

This module is the *executable contract*, not the production implementation:
the production aggregator is the native runtime owned by E17 and must reproduce
the semantics frozen here. Semantics ported (not vendored) from the review
package's reference oracle, extended with transformation change-counts,
timezone day grouping and formula versioning.

Frozen semantics (see packages/contracts/insight-events/README.md):

* Dedupe key: ``event_id``. Same ID + identical payload = idempotent replay
  (retry/sync). Same ID + different payload = conflict (ValueError).
* Recognition selection: latest ``selection_seq`` per capture wins; a retry
  *replaces* the word count, it never adds.
* Tombstones: any ``capture_deleted`` for a capture_id removes every
  attributable event of that capture, regardless of arrival order. Stale sync
  replays cannot resurrect deleted totals.
* WPM denominator: silence-inclusive captured seconds of takes that have a
  selected recognition ("recognized words per captured minute"), summed over
  the population (weighted totals, never an average of per-take rates).
* Generated/snippet words are counted from delivery events, separately from
  recognized speech words, and never enter the WPM or time proxy numerators.
* Time-saved proxy: typing estimate at a user baseline minus eligible capture
  seconds minus non-overlapping post-Stop wait. ``None`` when the baseline is
  unset or any wait is unknown; may be negative and is reported as such.

Pure stdlib; no tokenizer (word counts must be supplied precomputed by a
declared tokenizer), no storage, no network.
"""

from __future__ import annotations

import math
from datetime import datetime
from typing import Any
from zoneinfo import ZoneInfo, ZoneInfoNotFoundError

FORMULA_VERSION = 1
SUPPORTED_SCHEMA_VERSION = 1
DELIVERY_STATUSES = ("confirmed", "submitted_unconfirmed", "failed", "conflict", "cancelled")
CHANGE_KINDS = ("structural", "user", "dictionary", "snippet", "style")
TRANSFORMATION_KINDS = ("model_authoring", "snippet_expansion", "user_edit", "dictionary_substitution")


def nonnegative(value: Any, label: str) -> float:
    """Reject bools, non-numbers, non-finite and negative values."""
    if isinstance(value, bool) or not isinstance(value, (int, float)):
        raise ValueError(f"{label} must be a number")
    if not math.isfinite(value) or value < 0:
        raise ValueError(f"{label} must be finite and nonnegative")
    return float(value)


def latest_by_sequence(items: list[dict[str, Any]], field: str) -> dict[str, Any] | None:
    """Latest event by a monotone sequence field; equal seq + different payload = conflict."""
    seen: dict[int, dict[str, Any]] = {}
    for item in items:
        seq = item[field]
        if type(seq) is not int or seq < 0:
            raise ValueError(f"Invalid {field} sequence")
        if seq in seen and seen[seq] != item:
            raise ValueError(f"Conflicting {field} sequence values")
        seen[seq] = item
    return seen[max(seen)] if seen else None


def _dedupe_and_tombstone(events: list[dict[str, Any]]) -> dict[str, list[dict[str, Any]]]:
    """Idempotency dedupe by event_id, then drop tombstoned captures entirely."""
    unique: dict[str, dict[str, Any]] = {}
    for event in events:
        if event.get("schema_version") != SUPPORTED_SCHEMA_VERSION:
            raise ValueError("Unsupported event schema_version")
        id_ = event["event_id"]
        if id_ in unique and unique[id_] != event:
            raise ValueError("Same event ID has conflicting payloads")
        unique[id_] = event
    # Tombstones dominate stale replays; explicit restore is not part of v1.
    deleted = {e["capture_id"] for e in unique.values() if e["type"] == "capture_deleted"}
    grouped: dict[str, list[dict[str, Any]]] = {}
    for event in unique.values():
        if event["capture_id"] not in deleted:
            grouped.setdefault(event["capture_id"], []).append(event)
    return grouped


def _parse_utc(occurred_at: str) -> datetime:
    if occurred_at.endswith("Z"):
        occurred_at = occurred_at[:-1] + "+00:00"
    stamp = datetime.fromisoformat(occurred_at)
    if stamp.tzinfo is None:
        raise ValueError("occurred_at must carry an explicit UTC offset")
    return stamp.astimezone(ZoneInfo("UTC"))


def aggregate(events: list[dict[str, Any]], typing_wpm: float | None = None) -> dict[str, Any]:
    """Aggregate a local event log into the v1 metric contract.

    Raises ValueError on structurally impossible input (conflicting event IDs
    or sequences, impossible counts, non-positive sample rates) so bugs can
    never silently degrade into plausible numbers.
    """
    if typing_wpm is not None and nonnegative(typing_wpm, "typing_wpm") == 0:
        raise ValueError("typing_wpm must be positive")
    grouped = _dedupe_and_tombstone(events)

    count = words = raw_words = selected_count = incomplete = 0
    captured_seconds = eligible_seconds = wait_seconds = 0.0
    waits_known = True
    tokenizer_set: set[str] = set()
    delivery_counts: dict[str, int] = {k: 0 for k in DELIVERY_STATUSES}
    delivered_words = {k: 0 for k in DELIVERY_STATUSES}
    generated_words = {k: 0 for k in DELIVERY_STATUSES}
    change_counts = {k: 0 for k in CHANGE_KINDS}
    transformation_counts = {k: 0 for k in TRANSFORMATION_KINDS}

    for capture_id in sorted(grouped):
        items = grouped[capture_id]
        captures = [e for e in items if e["type"] == "capture_finalized"]
        if not captures:
            continue  # orphan metadata is not evidence of a captured take
        if len(captures) != 1:
            raise ValueError("Multiple canonical capture finalizations")
        capture = captures[0]
        sample_rate = nonnegative(capture["sample_rate"], "sample_rate")
        if sample_rate <= 0:
            raise ValueError("sample_rate must be positive")
        duration = nonnegative(capture["sample_count"], "sample_count") / sample_rate
        captured_seconds += duration
        count += 1
        incomplete += int(not capture["complete_audio"])

        selected = latest_by_sequence(
            [e for e in items if e["type"] == "recognition_selected"], "selection_seq"
        )
        if selected is not None:
            selected_count += 1
            lexical = nonnegative(selected["lexical_words"], "lexical_words")
            raw = nonnegative(selected["raw_words"], "raw_words")
            if lexical > raw:
                raise ValueError("Lexical count cannot exceed raw ASR count")
            words += int(lexical)
            raw_words += int(raw)
            eligible_seconds += duration
            tokenizer_set.add(selected["tokenizer"])
            wait = selected["post_stop_ready_ms"]
            if wait is None:
                waits_known = False
            else:
                wait_seconds += nonnegative(wait, "post_stop_ready_ms") / 1000

        # Transformation revisions: latest revision_seq per revision_id wins
        # (a refinement retry replaces, it does not add); distinct revision
        # IDs are distinct passes and both are counted. Change counts are
        # labelled "changes", never "corrected errors".
        revisions: dict[str, list[dict[str, Any]]] = {}
        for item in items:
            if item["type"] == "transformation_completed":
                revisions.setdefault(item["revision_id"], []).append(item)
        for revision_items in revisions.values():
            revision = latest_by_sequence(revision_items, "revision_seq")
            assert revision is not None
            transformation_counts[revision["transformation_kind"]] += 1
            for kind in CHANGE_KINDS:
                change_counts[kind] += int(
                    nonnegative(revision["change_counts"][kind], f"change_counts.{kind}")
                )

        deliveries: dict[str, list[dict[str, Any]]] = {}
        for item in items:
            if item["type"] == "delivery_recorded":
                deliveries.setdefault(item["delivery_id"], []).append(item)
        for same_delivery in deliveries.values():
            delivery = latest_by_sequence(same_delivery, "delivery_seq")
            assert delivery is not None
            status = delivery["status"]
            if status not in delivery_counts:
                raise ValueError("Unknown delivery acknowledgement")
            output = nonnegative(delivery["output_words"], "output_words")
            generated = nonnegative(delivery["generated_words"], "generated_words")
            if generated > output:
                raise ValueError("Generated words cannot exceed output words")
            delivery_counts[status] += 1
            delivered_words[status] += int(output)
            generated_words[status] += int(generated)

    comparable = len(tokenizer_set) == 1
    wpm = words * 60 / eligible_seconds if eligible_seconds > 0 and comparable else None
    proxy = (
        words * 60 / typing_wpm - eligible_seconds - wait_seconds
        if typing_wpm and waits_known and selected_count and comparable
        else None
    )
    return {
        "formula_version": FORMULA_VERSION,
        "unique_takes": count,
        "selected_recognitions": selected_count,
        "incomplete_captures": incomplete,
        "recognized_words": words,
        "raw_recognized_words": raw_words,
        "captured_seconds": captured_seconds,
        "eligible_capture_seconds": eligible_seconds,
        # Silence-inclusive capture-normalized rate; VAD speech-rate is a
        # distinct metric (E17 runtime) and is never silently substituted.
        "recognized_words_per_captured_minute": wpm,
        "tokenizers": sorted(tokenizer_set),
        "delivery_counts": delivery_counts,
        "output_words_by_status": delivered_words,
        "generated_words_by_status": generated_words,
        # Changes by type; NOT certified error corrections.
        "change_counts": change_counts,
        "transformation_counts": transformation_counts,
        "typing_time_comparison_seconds": proxy,
        "time_comparison_caveat": (
            "Estimate excludes unobserved correction time; generated words are "
            "not speech and never enter this estimate."
        ),
    }


def activity_by_day(
    events: list[dict[str, Any]], timezone_name: str
) -> dict[str, dict[str, Any]]:
    """Group attributable takes into local calendar days of a declared timezone.

    DST-aware via zoneinfo (a fallback wall-clock conversion would be a
    silent bug, so the function refuses to run without a tz database).
    """
    try:
        tz = ZoneInfo(timezone_name)
    except (ZoneInfoNotFoundError, ValueError) as exc:
        raise ValueError(f"Unknown reporting timezone: {timezone_name}") from exc
    grouped = _dedupe_and_tombstone(events)
    days: dict[str, dict[str, Any]] = {}
    for capture_id in sorted(grouped):
        items = grouped[capture_id]
        captures = [e for e in items if e["type"] == "capture_finalized"]
        if len(captures) != 1:
            if not captures:
                continue
            raise ValueError("Multiple canonical capture finalizations")
        capture = captures[0]
        sample_rate = nonnegative(capture["sample_rate"], "sample_rate")
        if sample_rate <= 0:
            raise ValueError("sample_rate must be positive")
        duration = nonnegative(capture["sample_count"], "sample_count") / sample_rate
        local_day = _parse_utc(capture["occurred_at"]).astimezone(tz).date().isoformat()
        bucket = days.setdefault(
            local_day, {"takes": 0, "captured_seconds": 0.0}
        )
        bucket["takes"] += 1
        bucket["captured_seconds"] += duration
    return days


def local_wall_time(occurred_at: str, timezone_name: str) -> tuple[int, int, int, float]:
    """(local hour, minute, second, utc_offset_hours) for DST-ambiguity proofs."""
    try:
        tz = ZoneInfo(timezone_name)
    except (ZoneInfoNotFoundError, ValueError) as exc:
        raise ValueError(f"Unknown reporting timezone: {timezone_name}") from exc
    local = _parse_utc(occurred_at).astimezone(tz)
    offset = local.utcoffset()
    assert offset is not None
    return local.hour, local.minute, local.second, offset.total_seconds() / 3600

File: apps/desktop/src/insights/insightMetrics.ts
import {
  CHANGE_KINDS,
  DELIVERY_STATUSES,
  INSIGHT_EVENT_SCHEMA_VERSION,
  isCaptureFinalized,
  isDeliveryRecorded,
  isRecognitionSelected,
  isTransformationCompleted,
  sameEventPayload,
  type ChangeKind,
  type DeliveryRecordedEvent,
  type DeliveryStatus,
  type InsightEvent,
  type TransformationCompletedEvent,
  type TransformationKind,
} from "./insightEvents";

/**
 * TypeScript port of the executable metric contract (E28) —
 * `tests/insight_metrics.py` is the oracle; this module reproduces its frozen
 * semantics for the desktop Insights surface without importing the Python:
 *
 * * Dedupe key: `event_id`. Same ID + equal payload = idempotent replay
 *   (retry/sync). Same ID + different payload = conflict (hard error).
 * * Recognition selection: latest `selection_seq` per capture wins; a retry
 *   *replaces* the word count, it never adds.
 * * Tombstones: any `capture_deleted` for a capture_id removes every
 *   attributable event of that capture, regardless of arrival order. Stale
 *   sync replays cannot resurrect deleted totals.
 * * WPM denominator: silence-inclusive captured seconds of takes that have a
 *   selected recognition ("recognized words per captured minute"), summed over
 *   the population (weighted totals, never an average of per-take rates).
 * * Generated/snippet words are counted from delivery events, separately from
 *   recognized speech words, and never enter the WPM or time proxy numerators.
 * * Time-saved proxy: typing estimate at a user baseline minus eligible
 *   capture seconds minus non-overlapping post-Stop wait. `null` when the
 *   baseline is unset or any wait is unknown; may be negative and is reported
 *   as such.
 *
 * Output field names keep the oracle's snake_case so an exported aggregate is
 * directly comparable against the reference implementation's JSON.
 */

export const FORMULA_VERSION = 1;

const TIME_COMPARISON_CAVEAT =
  "Estimate excludes unobserved correction time; generated words are not speech and never enter this estimate.";

export class InsightContractError extends Error {
  override readonly name = "InsightContractError";

  constructor(message: string) {
    super(message);
  }
}

/**
 * Reject non-finite and negative values. The Python oracle additionally
 * rejects bools and non-numbers before this check; here the event types carry
 * numbers, so the runtime guard has to catch what a cast can smuggle in:
 * NaN and the infinities (a plain `< 0` test alone lets NaN through).
 */
function nonnegative(value: number, label: string): number {
  if (!Number.isFinite(value) || value < 0) {
    throw new InsightContractError(`${label} must be finite and nonnegative`);
  }

  return value;
}

/**
 * Latest event by a monotone sequence field; equal seq + different payload =
 * conflict. Mirrors the oracle's `latest_by_sequence`, including the refusal
 * of fractional, negative or non-integer sequence values.
 */
function latestBySequence<T extends InsightEvent>(
  items: readonly T[],
  sequenceOf: (item: T) => number,
): T | undefined {
  const seen = new Map<number, T>();
  let highest: number | undefined;

  for (const item of items) {
    const seq = sequenceOf(item);

    if (!Number.isInteger(seq) || seq < 0) {
      throw new InsightContractError(`Invalid sequence value ${seq}`);
    }

    const existing = seen.get(seq);

    if (existing !== undefined && !sameEventPayload(existing, item)) {
      throw new InsightContractError("Conflicting sequence values");
    }

    seen.set(seq, item);

    if (highest === undefined || seq > highest) highest = seq;
  }

  if (highest === undefined) return undefined;

  const latest = seen.get(highest);

  return latest;
}

/**
 * Idempotency dedupe by event_id, then drop tombstoned captures entirely.
 * Tombstones dominate stale replays; explicit restore is not part of v1.
 */
function dedupeAndTombstone(events: readonly InsightEvent[]): Map<string, InsightEvent[]> {
  const unique = new Map<string, InsightEvent>();

  for (const event of events) {
    if (event.schema_version !== INSIGHT_EVENT_SCHEMA_VERSION) {
      throw new InsightContractError("Unsupported event schema_version");
    }

    const existing = unique.get(event.event_id);

    if (existing !== undefined && !sameEventPayload(existing, event)) {
      throw new InsightContractError("Same event ID has conflicting payloads");
    }

    unique.set(event.event_id, event);
  }

  const deleted = new Set<string>();

  for (const event of unique.values()) {
    if (event.type === "capture_deleted") deleted.add(event.capture_id);
  }

  const grouped = new Map<string, InsightEvent[]>();

  for (const event of unique.values()) {
    if (deleted.has(event.capture_id)) continue;

    const items = grouped.get(event.capture_id);

    if (items === undefined) grouped.set(event.capture_id, [event]);
    else items.push(event);
  }

  return grouped;
}

interface CanonicalCapture {
  readonly sampleCount: number;
  readonly sampleRate: number;
  readonly completeAudio: boolean;
  readonly occurredAt: string;
}

/**
 * The one canonical finalization of a take, under the oracle's exact rules:
 * undefined when the capture has no finalization event (orphan metadata is
 * not evidence of a captured take), an error when two claim to be canonical.
 */
function canonicalCapture(items: readonly InsightEvent[]): CanonicalCapture | undefined {
  const captures = items.filter(isCaptureFinalized);

  if (captures.length === 0) return undefined;

  if (captures.length !== 1) {
    throw new InsightContractError("Multiple canonical capture finalizations");
  }

  const capture = captures[0];

  if (capture === undefined) {
    throw new InsightContractError("Multiple canonical capture finalizations");
  }

  return {
    sampleCount: capture.sample_count,
    sampleRate: capture.sample_rate,
    completeAudio: capture.complete_audio,
    occurredAt: capture.occurred_at,
  };
}

/** Capture duration in seconds; refuses impossible rates and counts. */
function captureSeconds(capture: CanonicalCapture): number {
  const sampleRate = nonnegative(capture.sampleRate, "sample_rate");

  if (sampleRate <= 0) throw new InsightContractError("sample_rate must be positive");

  return nonnegative(capture.sampleCount, "sample_count") / sampleRate;
}

export interface InsightAggregate {
  readonly formula_version: typeof FORMULA_VERSION;
  readonly unique_takes: number;
  readonly selected_recognitions: number;
  readonly incomplete_captures: number;
  readonly recognized_words: number;
  readonly raw_recognized_words: number;
  readonly captured_seconds: number;
  readonly eligible_capture_seconds: number;
  /** Silence-inclusive capture-normalized rate; null rather than corrupted. */
  readonly recognized_words_per_captured_minute: number | null;
  readonly tokenizers: readonly string[];
  readonly delivery_counts: Readonly<Record<DeliveryStatus, number>>;
  readonly output_words_by_status: Readonly<Record<DeliveryStatus, number>>;
  readonly generated_words_by_status: Readonly<Record<DeliveryStatus, number>>;
  /** Changes by type; NOT certified error corrections. */
  readonly change_counts: Readonly<Record<ChangeKind, number>>;
  readonly transformation_counts: Readonly<Record<TransformationKind, number>>;
  readonly typing_time_comparison_seconds: number | null;
  readonly time_comparison_caveat: string;
}

function zeroedStatusCounts(): Record<DeliveryStatus, number> {
  return {
    confirmed: 0,
    submitted_unconfirmed: 0,
    failed: 0,
    conflict: 0,
    cancelled: 0,
  };
}

function zeroedChangeCounts(): Record<ChangeKind, number> {
  return { structural: 0, user: 0, dictionary: 0, snippet: 0, style: 0 };
}

function zeroedTransformationCounts(): Record<TransformationKind, number> {
  return {
    model_authoring: 0,
    snippet_expansion: 0,
    user_edit: 0,
    dictionary_substitution: 0,
  };
}

const DELIVERY_STATUS_SET: ReadonlySet<string> = new Set(DELIVERY_STATUSES);

/**
 * Aggregate a local event log into the v1 metric contract. Throws
 * InsightContractError on structurally impossible input (conflicting event
 * IDs or sequences, impossible counts, non-positive sample rates) so bugs can
 * never silently degrade into plausible numbers.
 */
export function aggregate(
  events: readonly InsightEvent[],
  typingWpm?: number | null,
): InsightAggregate {
  if (typingWpm !== undefined && typingWpm !== null && nonnegative(typingWpm, "typing_wpm") === 0) {
    throw new InsightContractError("typing_wpm must be positive");
  }

  const grouped = dedupeAndTombstone(events);

  let count = 0;
  let words = 0;
  let rawWords = 0;
  let selectedCount = 0;
  let incomplete = 0;
  let capturedSeconds = 0;
  let eligibleSeconds = 0;
  let waitSeconds = 0;
  let waitsKnown = true;
  const tokenizers = new Set<string>();
  const deliveryCounts = zeroedStatusCounts();
  const deliveredWords = zeroedStatusCounts();
  const generatedWords = zeroedStatusCounts();
  const changeCounts = zeroedChangeCounts();
  const transformationCounts = zeroedTransformationCounts();

  for (const captureId of [...grouped.keys()].sort()) {
    const items = grouped.get(captureId) ?? [];
    const capture = canonicalCapture(items);

    // Orphan metadata is not evidence of a captured take.
    if (capture === undefined) continue;

    const duration = captureSeconds(capture);

    capturedSeconds += duration;
    count += 1;
    incomplete += capture.completeAudio ? 0 : 1;

    const selected = latestBySequence(
      items.filter(isRecognitionSelected),
      (event) => event.selection_seq,
    );

    if (selected !== undefined) {
      selectedCount += 1;

      const lexical = nonnegative(selected.lexical_words, "lexical_words");
      const raw = nonnegative(selected.raw_words, "raw_words");

      if (lexical > raw) {
        throw new InsightContractError("Lexical count cannot exceed raw ASR count");
      }

      words += Math.trunc(lexical);
      rawWords += Math.trunc(raw);
      eligibleSeconds += duration;
      tokenizers.add(selected.tokenizer);

      if (selected.post_stop_ready_ms === null) {
        waitsKnown = false;
      } else {
        waitSeconds += nonnegative(selected.post_stop_ready_ms, "post_stop_ready_ms") / 1000;
      }
    }

    // Transformation revisions: latest revision_seq per revision_id wins
    // (a refinement retry replaces, it does not add); distinct revision IDs
    // are distinct passes and both are counted. Change counts are labelled
    // "changes", never "corrected errors".
    const revisions = new Map<string, TransformationCompletedEvent[]>();

    for (const item of items) {
      if (!isTransformationCompleted(item)) continue;

      const list = revisions.get(item.revision_id);

      if (list === undefined) revisions.set(item.revision_id, [item]);
      else list.push(item);
    }

    for (const revisionItems of revisions.values()) {
      const revision = latestBySequence(revisionItems, (event) => event.revision_seq);

      if (revision === undefined) continue;

      transformationCounts[revision.transformation_kind] += 1;

      for (const kind of CHANGE_KINDS) {
        changeCounts[kind] += Math.trunc(
          nonnegative(revision.change_counts[kind], `change_counts.${kind}`),
        );
      }
    }

    const deliveries = new Map<string, DeliveryRecordedEvent[]>();

    for (const item of items) {
      if (!isDeliveryRecorded(item)) continue;

      const list = deliveries.get(item.delivery_id);

      if (list === undefined) deliveries.set(item.delivery_id, [item]);
      else list.push(item);
    }

    for (const deliveryItems of deliveries.values()) {
      const delivery = latestBySequence(deliveryItems, (event) => event.delivery_seq);

      if (delivery === undefined) continue;

      if (!DELIVERY_STATUS_SET.has(delivery.status)) {
        throw new InsightContractError("Unknown delivery acknowledgement");
      }

      const output = nonnegative(delivery.output_words, "output_words");
      const generated = nonnegative(delivery.generated_words, "generated_words");

      if (generated > output) {
        throw new InsightContractError("Generated words cannot exceed output words");
      }

      deliveryCounts[delivery.status] += 1;
      deliveredWords[delivery.status] += Math.trunc(output);
      generatedWords[delivery.status] += Math.trunc(generated);
    }
  }

  const comparable = tokenizers.size === 1;
  const wpm = eligibleSeconds > 0 && comparable ? (words * 60) / eligibleSeconds : null;

  const proxy =
    typingWpm !== undefined && typingWpm !== null && waitsKnown && selectedCount > 0 && comparable
      ? (words * 60) / typingWpm - eligibleSeconds - waitSeconds
      : null;

  return {
    formula_version: FORMULA_VERSION,
    unique_takes: count,
    selected_recognitions: selectedCount,
    incomplete_captures: incomplete,
    recognized_words: words,
    raw_recognized_words: rawWords,
    captured_seconds: capturedSeconds,
    eligible_capture_seconds: eligibleSeconds,
    recognized_words_per_captured_minute: wpm,
    tokenizers: [...tokenizers].sort(),
    delivery_counts: deliveryCounts,
    output_words_by_status: deliveredWords,
    generated_words_by_status: generatedWords,
    change_counts: changeCounts,
    transformation_counts: transformationCounts,
    typing_time_comparison_seconds: proxy,
    time_comparison_caveat: TIME_COMPARISON_CAVEAT,
  };
}

export interface DayActivity {
  readonly takes: number;
  readonly captured_seconds: number;
}

/** Selection-derived quality counts, on the same dedupe/tombstone rules. */
export interface SelectionStats {
  /** Non-deleted captures with at least one selected recognition. */
  readonly capturesWithSelection: number;
  /** Selections a newer selection replaced — retries, not extra takes. */
  readonly supersededSelections: number;
  /** Captures whose selected recognition recognized zero words. */
  readonly emptySelections: number;
}

export function selectionStats(events: readonly InsightEvent[]): SelectionStats {
  const grouped = dedupeAndTombstone(events);
  let capturesWithSelection = 0;
  let supersededSelections = 0;
  let emptySelections = 0;

  for (const items of grouped.values()) {
    const selections = items.filter(isRecognitionSelected);

    if (selections.length === 0) continue;

    capturesWithSelection += 1;
    supersededSelections += selections.length - 1;

    const selected = latestBySequence(selections, (event) => event.selection_seq);

    if (selected !== undefined && selected.lexical_words === 0) emptySelections += 1;
  }

  return { capturesWithSelection, supersededSelections, emptySelections };
}

/** The local calendar-day key of an instant, in the declared timezone. */
export function localDayKey(ms: number, timezoneName: string): string {
  return localDayFormatter(timezoneName).format(ms);
}

/**
 * Group attributable takes into local calendar days of a declared timezone.
 * DST-aware through the host's IANA timezone database — a naive wall-clock
 * conversion would silently misplace the fallback hour, so an unknown
 * timezone refuses to run rather than guessing.
 */
export function activityByDay(
  events: readonly InsightEvent[],
  timezoneName: string,
): Map<string, DayActivity> {
  const dayFormatter = localDayFormatter(timezoneName);
  const grouped = dedupeAndTombstone(events);
  const days = new Map<string, DayBucket>();

  for (const captureId of [...grouped.keys()].sort()) {
    const items = grouped.get(captureId) ?? [];
    const capture = canonicalCapture(items);

    if (capture === undefined) continue;

    const duration = captureSeconds(capture);
    const localDay = dayFormatter.format(parseInsightTimestamp(capture.occurredAt));
    const bucket = days.get(localDay) ?? { takes: 0, captured_seconds: 0 };

    bucket.takes += 1;
    bucket.captured_seconds += duration;
    days.set(localDay, bucket);
  }

  return days;
}

interface DayBucket {
  takes: number;
  captured_seconds: number;
}

export interface LocalWallTime {
  readonly hour: number;
  readonly minute: number;
  readonly second: number;
  readonly utc_offset_hours: number;
}

/**
 * (local hour, minute, second, UTC offset in hours) for DST-ambiguity proofs:
 * during a fallback the same wall time recurs under two offsets, and the
 * offset — not the wall clock — disambiguates them.
 */
export function localWallTime(occurredAt: string, timezoneName: string): LocalWallTime {
  const formatter = localWallFormatter(timezoneName);
  const parts = formatter.formatToParts(parseInsightTimestamp(occurredAt));
  const values = new Map<string, string>();

  for (const part of parts) {
    if (part.type !== "literal") values.set(part.type, part.value);
  }

  const hour = Number.parseInt(values.get("hour") ?? "", 10);
  const minute = Number.parseInt(values.get("minute") ?? "", 10);
  const second = Number.parseInt(values.get("second") ?? "", 10);

  if (![hour, minute, second].every((part) => Number.isInteger(part))) {
    throw new InsightContractError(`could not read local wall time in ${timezoneName}`);
  }

  return {
    hour,
    minute,
    second,
    utc_offset_hours: parseGmtOffset(values.get("timeZoneName") ?? ""),
  };
}

/** Parse a "GMT+02:00"/"GMT-1"/"GMT" long-offset name into signed hours. */
function parseGmtOffset(text: string): number {
  const match = /^GMT(?:([+-])(\d{1,2})(?::(\d{2}))?)?$/.exec(text);

  if (match === null) {
    throw new InsightContractError(`could not read a UTC offset from ${text}`);
  }

  if (match[1] === undefined) return 0;

  const hours = Number.parseInt(match[2] ?? "0", 10);
  const minutes = match[3] === undefined ? 0 : Number.parseInt(match[3], 10);
  const magnitude = hours + minutes / 60;

  return match[1] === "-" ? -magnitude : magnitude;
}

const INSIGHT_TIMESTAMP_PATTERN =
  /^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}(\.[0-9]+)?(Z|[+-][0-9]{2}:[0-9]{2})$/;

/** Parse an occurred_at stamp; a naive timestamp without an offset refuses. */
export function parseInsightTimestamp(occurredAt: string): number {
  if (!INSIGHT_TIMESTAMP_PATTERN.test(occurredAt)) {
    throw new InsightContractError("occurred_at must carry an explicit UTC offset");
  }

  const ms = Date.parse(occurredAt);

  if (Number.isNaN(ms)) {
    throw new InsightContractError("occurred_at must carry an explicit UTC offset");
  }

  return ms;
}

const dayFormatters = new Map<string, Intl.DateTimeFormat>();

const wallFormatters = new Map<string, Intl.DateTimeFormat>();

function localDayFormatter(timezoneName: string): Intl.DateTimeFormat {
  let formatter = dayFormatters.get(timezoneName);

  if (formatter === undefined) {
    formatter = resolveTimezone(timezoneName, () => buildLocalDayFormatter(timezoneName));
    dayFormatters.set(timezoneName, formatter);
  }

  return formatter;
}

function localWallFormatter(timezoneName: string): Intl.DateTimeFormat {
  let formatter = wallFormatters.get(timezoneName);

  if (formatter === undefined) {
    formatter = resolveTimezone(timezoneName, () => buildLocalWallFormatter(timezoneName));
    wallFormatters.set(timezoneName, formatter);
  }

  return formatter;
}

function buildLocalDayFormatter(timezoneName: string): Intl.DateTimeFormat {
  return new Intl.DateTimeFormat("en-CA", {
    timeZone: timezoneName,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  });
}

function buildLocalWallFormatter(timezoneName: string): Intl.DateTimeFormat {
  return new Intl.DateTimeFormat("en-GB", {
    timeZone: timezoneName,
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    hourCycle: "h23",
    timeZoneName: "longOffset",
  });
}

function resolveTimezone(
  timezoneName: string,
  build: () => Intl.DateTimeFormat,
): Intl.DateTimeFormat {
  try {
    return build();
  } catch {
    throw new InsightContractError(`Unknown reporting timezone: ${timezoneName}`);
  }
}
